net/dns/resolver/forwarder.go

```diff
@@ -701,7 +701,7 @@ func (f *forwarder) sendTCP(ctx context.Context, fq *forwardQuery, rr resolverAn
 	ctx, cancel := context.WithTimeout(ctx, tcpQueryTimeout)
 	defer cancel()
 
-	conn, err := f.dialer.SystemDial(ctx, tcpFam, ipp.String())
+	conn, err := f.dialer.UserDial(ctx, tcpFam, ipp.String())
```

Comment:
@andrew-d This in particular I have trouble evaluating.

Reply:
I spent a bunch of time looking at this, and I'm not 100% sure this is correct. Specifically: `SystemDial` avoids the Tailscale overlay network in all cases, including routes advertised with `--advertise-routes`. `UserDial` dials out as if the system was in `tun` mode, routing through Tailscale (though with the caveat that it doesn't currently resolve split DNS correctly).

So, if the tailnet has a node A with `--advertise-routes=1.1.1.0/24`, and a split DNS entry saying "send queries for `example.com` to `1.1.1.1`", then previously node B would send packets directly to `1.1.1.1` when making DNS queries, and with this change, queries from node B would go through node A.

I'm going to attempt to write some tests to verify this behaviour.

Reply:
I agree this does create a divergence in behavior but it is the exact desired behavior in my case. In my setup, I run the `tailscaled` in k8s which it advertise a route to its private IP range and setup SplitDNS to send `.svc.cluster.local` queries to k8s's CoreDNS server.

In `tun` mode, these domains can be resolved correctly because the query are sent to that k8s internal DNS server. However, if we do `SystemDial` here, we won't be able to connect to the server because it's on a private route advertised by other `tailscaled`.

Reply:
Yeah, the thing I want to determine is if there's cases where this would break connectivity for a user, including users that aren't using k8s. One of the cases, for example, is if you try to query `example.com` from node A in my example above. `SystemDial` properly avoids routing packets back over Tailscale, so the query would go to the `1.1.1.1` IP directly from node A. However, if node A is using `UserDial` to make that connection, it might try to route packets back over Tailscale to get to `1.1.1.1`, which I want to test.

Reply:
I'll ask that this PR change this line back to SystemDial. I understand this means that it won't immediately fix the issue you're seeing.

We plan to release 1.58 this coming week. We would submit this PR after the 1.58 release branch is made. This PR would first appear in 1.60.

We need to implement the functionality to switch from SystemDial to UserDial in a way that we can disable from the coordination server, switching back to UserDial if needed. We cannot send 1.60 out having maade such a substantial change in behavior and then need an emergency 1.60.1 release if this introduces significant regressions for many users.

Reply:
@DentonGentry is this something that could be put behind an "experimental" environment variable feature gate?

Reply:
okay. I understand your concern and has reverted for now. Left a pointer at a line. If we're open to having a flag instead, I'm happy to take on that work as well.